src/channels/minesweeper/utils.ts

```diff
@@ -0,0 +1,86 @@
+import { TileData } from './Tile';
+import { GRID_COLUMNS, GRID_ROWS, REVEAL_CHANCE_DECAY } from './constants';
+
+function random(min: number, max: number) {
+	return Math.floor(Math.random() * (max - min)) + min;
+}
+
+export function randomFromArray<T>(arr: T[]) {
+	const randomIndex = random(0, arr.length);
+	return arr[randomIndex];
+}
+
+export function splitTileIndex(indexStr: string) {
+	return indexStr.split(':').map((index) => Number(index));
+}
+
+export function isTileInbounds(rowIndex: number, tileIndex: number) {
+	return rowIndex >= 0 && rowIndex < GRID_ROWS && tileIndex >= 0 && tileIndex < GRID_COLUMNS;
+}
+
+function getAdjacentTiles(rowIndex: number, tileIndex: number) {
+	return [
+		[rowIndex - 1, tileIndex],
+		[rowIndex, tileIndex - 1],
+		[rowIndex, tileIndex + 1],
+		[rowIndex + 1, tileIndex],
+	];
+}
+
+function getSurroundingTiles(rowIndex: number, tileIndex: number) {
+	return [
+		[rowIndex - 1, tileIndex - 1],
+		[rowIndex - 1, tileIndex],
+		[rowIndex - 1, tileIndex + 1],
+		[rowIndex, tileIndex - 1],
+		[rowIndex, tileIndex + 1],
+		[rowIndex + 1, tileIndex - 1],
+		[rowIndex + 1, tileIndex],
+		[rowIndex + 1, tileIndex + 1],
+	];
+}
+
+export function getMineCount(grid: TileData[][], tileId: string) {
+	const [rowIndex, tileIndex] = splitTileIndex(tileId);
+	const surroundingTiles = getSurroundingTiles(rowIndex, tileIndex);
+	const surroundingMineCount = surroundingTiles.reduce((mineCount, [r, t]) => {
+		if (!isTileInbounds(r, t)) {
+			return mineCount;
+		}
+		return mineCount + (grid[r][t].isMine ? 1 : 0);
+	}, 0);
+	return surroundingMineCount;
+}
+
+export function createTileCluster(grid: TileData[][], startingTileId: string, revealChange: number) {
+	const visitedTiles: string[] = [];
+	const [rowIndex, tileIndex] = splitTileIndex(startingTileId);
+
+	function visitTile(rowIndex: number, tileIndex: number, chance = revealChange) {
+		const tile = grid[rowIndex]?.[tileIndex] as TileData | undefined;
+		// stop tile from being revealed (and cascading further)
+		// when the following conditions are met
+		if (
+			// not a valid tile
+			!tile ||
+			// the tile is a mine
+			tile.isMine ||
+			// has already been visisted
+			visitedTiles.includes(tile.id) ||
+			// or does not meet the reveal chance
+			// note: we always want the first tile chosen to be revealed
+			(visitedTiles.length > 0 && Math.random() > chance)
+		) {
+			return;
+		}
+		visitedTiles.push(tile.id);
+		const adjacentTiles = getAdjacentTiles(rowIndex, tileIndex);
+		adjacentTiles.forEach(([ajacentRowIndex, adjacentTileIndex]) =>
+			visitTile(ajacentRowIndex, adjacentTileIndex, chance - REVEAL_CHANCE_DECAY),
+		);
+	}
+
+	visitTile(rowIndex, tileIndex);
+
+	return visitedTiles;
+}
```

Comment:
Is there a particular reason this function uses a random walk and decay, rather than behaving like real Minesweeper, where a tile press with a number just reveals that square, but pressing an empty tile with no adjacent tiles will fully cascade until an entire region surrounded by numbers is filled? The random behavior leads to some revealed tiles that look unlike anything the game would ever actually produce.

From what I've been able to tell, you have most of the infrastructure already setup to do that, and should only require a few small changes to make work in that way?

Reply:
By using a decay I can roughly control the size of the revealed cluster. This is nice because we can set a lower decay for larger donations so a large area is more likely to be uncovered. To achieve the same functionality with the original game's "algorithm", we would need to know how much area is revealed by any given tile ahead of time, which is slightly beyond my skill set to figure out and felt like overkill anyway.

Reply:
Hmm. That is a fair enough point. Would you consider a middle ground, where the reveal function behaves like normal Minesweeper, but counts how many tiles were revealed by a particular selection, and ensures that if a large donation comes in, it'll keep running the reveal function until some threshold of revealed tiles is met (or I guess there are no more tiles to reveal)? That way you still have the large donations having a big impact, but it still looks like a real game, and small donations still have a chance to just be lucky with random chance, which I'd say is fine, if not preferred.

If you disagree, or aren't interested in implementing something like that yourself, I'll probably still merge this as-is and consider making the changes myself if I can get it to a place that I think works and if you're okay with such a modification being done.

Reply:
@VodBox I'm totally fine with the modification proposed and I think it's actually a great idea. Sorry for the late reply, the last couple weeks have been extra busy for me.

Reply:
@VodBox I can't believe I forgot about this for so long 🤦 I have *finally* implemented this suggested change and would appreciate if you gave it a try locally and tell me what you think. Feel free to play around with the threshold numbers in `constants.ts` if it doesn't feel right.